File: FBSimulatorControl/Framebuffer/FBSimulatorImage.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorImage.h"

#import <objc/runtime.h>

#import <CoreImage/CoreImage.h>

#import <FBControlCore/FBControlCore.h>

#import "FBFramebuffer.h"
#import "FBSimulatorError.h"
#import "FBSurfaceImageGenerator.h"

@interface FBSimulatorImage ()

@property (nonatomic, readonly, strong) id<FBControlCoreLogger> logger;
@property (nonatomic, readonly, strong) dispatch_queue_t writeQueue;
@property (nonatomic, readonly, strong) FBSurfaceImageGenerator *imageGenerator;
@property (nonatomic, readonly, strong) FBFramebuffer *framebuffer;
@property (nonatomic, readwrite, strong) NSUUID *consumerUUID;

@end

@implementation FBSimulatorImage

#pragma mark Initializers

+ (dispatch_queue_t)writeQueue
{
  return dispatch_queue_create("com.facebook.FBSimulatorControl.framebuffer.image", DISPATCH_QUEUE_SERIAL);
}

+ (instancetype)imageWithFramebuffer:(FBFramebuffer *)framebuffer logger:(id<FBControlCoreLogger>)logger
{
  return [[FBSimulatorImage alloc] initWithFramebuffer:framebuffer logger:logger];
}

- (instancetype)initWithFramebuffer:(FBFramebuffer *)framebuffer logger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _framebuffer = framebuffer;
  _logger = logger;
  _consumerUUID = [NSUUID UUID];
  _writeQueue = FBSimulatorImage.writeQueue;
  _imageGenerator = [FBSurfaceImageGenerator imageGeneratorWithScale:NSDecimalNumber.one purpose:@"simulator_image" logger:self.logger];

  return self;
}

#pragma mark FBSimulatorImage

- (nullable CGImageRef)image
{
  if (![self.framebuffer isConsumerAttached:self.imageGenerator]) {
    [self.logger log:[NSString stringWithFormat:@"Image Generator %@ not attached, attaching", self.imageGenerator]];
    IOSurface *surface = [self.framebuffer attachConsumer:self.imageGenerator onQueue:self.writeQueue];
    if (surface) {
      [self.logger log:[NSString stringWithFormat:@"Surface %@ immediately available, adding to Image Generator %@", surface, self.imageGenerator]];
      [self.imageGenerator didChangeIOSurface:surface];
    } else {
      [self.logger log:@"Surface for ImageGenerator not immedately available"];
    }
  }

  CGImageRef image = self.imageGenerator.image;
  if (image) {
    return image;
  }
  return self.imageGenerator.image;
}

- (nullable NSData *)jpegImageDataWithError:(NSError **)error
{
  return [FBSimulatorImage jpegImageDataFromImage:self.image error:error];
}

- (nullable NSData *)pngImageDataWithError:(NSError **)error
{
  return [FBSimulatorImage pngImageDataFromImage:self.image error:error];
}

#pragma mark Private

+ (nullable NSData *)jpegImageDataFromImage:(nullable CGImageRef)image error:(NSError **)error
{
  return [self imageDataFromImage:image type:kUTTypeJPEG error:error];
}

+ (nullable NSData *)pngImageDataFromImage:(nullable CGImageRef)image error:(NSError **)error
{
  return [self imageDataFromImage:image type:kUTTypePNG error:error];
}

+ (nullable NSData *)imageDataFromImage:(nullable CGImageRef)image type:(CFStringRef)type error:(NSError **)error
{
  if (!image) {
    return [[FBSimulatorError
             describe:@"No Image available to encode"]
            fail:error];
  }

  NSMutableData *data = [NSMutableData data];
  CGImageDestinationRef destination = CGImageDestinationCreateWithData(
    (CFMutableDataRef) data,
    type,
    1,
    NULL
  );
  CGImageDestinationAddImage(destination, image, NULL);
  if (!CGImageDestinationFinalize(destination)) {
    CFRelease(destination);
    return [[FBSimulatorError
             describe:@"Could not finalize the creation of the Image"]
            fail:error];
  }
  CFRelease(destination);
  return data;
}

@end

File: FBSimulatorControl/Framebuffer/FBSimulatorImage.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import CoreImage
@preconcurrency import FBControlCore
import Foundation
@preconcurrency import IOSurface
import ImageIO
import UniformTypeIdentifiers

/// The ways image encoding can fail, as data rather than assembled strings.
public enum FBSimulatorImageError: Error, LocalizedError {
  case noImage
  case destinationCreationFailed
  case finalizationFailed

  public var errorDescription: String? {
    switch self {
    case .noImage:
      return "No Image available to encode"
    case .destinationCreationFailed:
      return "Could not create image destination"
    case .finalizationFailed:
      return "Could not finalize the creation of the Image"
    }
  }
}

public actor FBSimulatorImage {

  // MARK: - Properties

  private let logger: (any FBControlCoreLogger)?
  private let imageGenerator: FBSurfaceImageGenerator
  private let framebuffer: FBFramebuffer
  private var attachment: FBFramebufferAttachment?
  private var eventTask: Task<Void, Never>?

  // MARK: - Initializers

  public static func image(with framebuffer: FBFramebuffer, logger: (any FBControlCoreLogger)?) -> FBSimulatorImage {
    FBSimulatorImage(framebuffer: framebuffer, logger: logger)
  }

  init(framebuffer: FBFramebuffer, logger: (any FBControlCoreLogger)?) {
    self.framebuffer = framebuffer
    self.logger = logger
    self.imageGenerator = FBSurfaceImageGenerator(scale: NSDecimalNumber.one, purpose: "simulator_image", logger: logger)
  }

  deinit {
    // Finishing the stream ends the event task, which releases the attachment.
    attachment?.cancel()
    eventTask?.cancel()
  }

  // MARK: - Public Methods

  public func image() throws -> CGImage? {
    try attachIfNeeded()
    return imageGenerator.image()
  }

  public func jpegImageData() throws -> Data {
    try FBSimulatorImage.jpegImageData(from: image())
  }

  public func pngImageData() throws -> Data {
    try FBSimulatorImage.pngImageData(from: image())
  }

  // MARK: - Private

  /// One-time lazy attach; actor isolation makes this attach-exactly-once regardless of caller
  /// threading. The generator's surface is seeded synchronously from `initialSurface`, then kept
  /// current by a task consuming the attachment's ordered event stream.
  private func attachIfNeeded() throws {
    guard attachment == nil else {
      return
    }
    logger?.log("Image Generator \(imageGenerator) not attached, attaching")
    let attachment = try framebuffer.attach()
    self.attachment = attachment
    if let surface = attachment.initialSurface {
      logger?.log("Surface \(surface) immediately available, adding to Image Generator \(imageGenerator)")
      imageGenerator.updateSurface(surface)
    } else {
      logger?.log("Surface for ImageGenerator not immediately available")
    }
    eventTask = Task { [weak self] in
      for await event in attachment.events {
        guard case let .surfaceChanged(surface) = event else {
          continue
        }
        await self?.applySurface(surface)
      }
    }
  }

  private func applySurface(_ surface: IOSurface?) {
    imageGenerator.updateSurface(surface)
  }

  // MARK: - Encoding

  private static func jpegImageData(from image: CGImage?) throws -> Data {
    try imageData(from: image, type: .jpeg)
  }

  private static func pngImageData(from image: CGImage?) throws -> Data {
    try imageData(from: image, type: .png)
  }

  private static func imageData(from image: CGImage?, type: UTType) throws -> Data {
    guard let image else {
      throw FBSimulatorImageError.noImage
    }

    let data = NSMutableData()
    guard let destination = CGImageDestinationCreateWithData(data as CFMutableData, type.identifier as CFString, 1, nil) else {
      throw FBSimulatorImageError.destinationCreationFailed
    }
    CGImageDestinationAddImage(destination, image, nil)
    if !CGImageDestinationFinalize(destination) {
      throw FBSimulatorImageError.finalizationFailed
    }
    return data as Data
  }
}
